packages/trace-viewer/src/ui/attachmentsTab.tsx

```diff
@@ -20,9 +20,58 @@ import { ImageDiffView } from '@web/shared/imageDiffView';
 import type { MultiTraceModel } from './modelUtil';
 import { PlaceholderPanel } from './placeholderPanel';
 import type { AfterActionTraceEventAttachment } from '@trace/trace';
+import { CodeMirrorWrapper } from '@web/components/codeMirrorWrapper';
+import { isTextualMimeType } from '@isomorphic/mimeType';
+import { Expandable } from '@web/components/expandable';
 
 type Attachment = AfterActionTraceEventAttachment & { traceUrl: string };
 
+type TextAttachmentProps = {
+  attachmentText: string;
+  label?: string;
+};
+
+const TextAttachment: React.FunctionComponent<TextAttachmentProps> = ({ attachmentText, label }) => {
+  return <div aria-label={label}>
+    {
+      attachmentText ?
+        <CodeMirrorWrapper text={attachmentText} readOnly wrapLines={false}></CodeMirrorWrapper> :
+        <div><i>Loading...</i></div>
+    }
+  </div>;
+};
+
+const ExpandableAttachment: React.FunctionComponent<{ attachment: Attachment }> = ({ attachment }) => {
+  const [expanded, setExpanded] = React.useState(false);
+  const [loaded, setLoaded] = React.useState(false);
+  const [attachmentText, setAttachmentText] = React.useState<string | null>(null);
+
+  React.useMemo(() => {
+    if (expanded && !loaded) {
+      fetch(attachmentURL(attachment)).then(response => response.text()).then(text => {
```

Comment:
Sounds like you'll be fetching the attachment regardless of its isTextualMimeType value. Do you want to only render Expandable when there is a potential preview?

Reply:
you're absolutely right, missed that while refactoring, thanks.

Should be fixed now: [342cace](https://github.com/microsoft/playwright/pull/31215/commits/342cace03d110f37acfb8aa85f9939d1d8fbfda0) - I also took the chance to simplify the component